docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+        var compilationSource = context.CompilationProvider;
+
+        context.RegisterArtifactOutput(compilationSource, (nspc, item) =>
+        {
+            nspc.AddFile("interop.js", (stream) =>
+            {
+                using var writer = new StreamWriter(stream);
+                writer.WriteLine("//output");
+            });
+        });
+    }
+}
+```
+
+Note how the call to `AddFile` takes a callback that provides a stream. This allows the host of the generator driver to control how the stream is created and what kind of stream is created. For instance in the compiler it could be a `FileStream` that writes directly to disk, but in unit tests it could be a `MemoryStream` that can be inspected as part of the test. It also takes a `hintName`, the same way that source files do, rather than an explicit file name or path. This is explained in more detail below.
+
+In order to control whether this output is used or not, a new entry is added to the `IncrementalGeneratorOutputKind` enum:
+
+```diff
+[Flags]
+public enum IncrementalGeneratorOutputKind
+{
+    None = 0,
+    Source = 0b1,
+    PostInit = 0b10,
++    Artifact = 0b100
+}
+```
+
+A host such as the IDE may disable non-source outputs when constructing the Generator Driver.
+
+In terms of the compilation and inputs that an Artifact can be based on, artifact outputs are treated the same as source outputs; all generation is based on the state of the compilation *before* any other generation has happened, and an artifact can't use the output of another generator (outside of itself) to generate artifacts.
+
+### File names and output locations
+
+The call to `AddFile` takes a `hintName` for the output file rather than an explicit file name or path. This allows the host of the generator driver the freedom to decide where precisely the file should go (if it is indeed going to e.g. the disk). Here we propose a set of rules that the command line compiler will follow. Other hosts may choose a different scheme.
+
+The compiler will take a new command line option `/generatedartifactsout:` that will specify the _default_ location of any generated artifacts. By default, using the compiler MSBuild targets, this will be set to the `$(OutDir)` property (i.e. `/bin/`). Note that this is different to `OutputAssembly` and `GeneratedFilesOutputPath` which typically use the `$(IntermediateOutputPath)` (i.e. `/obj/`). This ensures that artifacts are written to their final location a single time, rather than being written to an intermediate directory and having to be copied, optimizing the disk usage.
+
+By default the supplied `hintName` will be used as the filename. We already have conditions in place for source texts to prevent hint names from being invalid filenames, and we will reuse this logic for artifact names. If two generators attempt to write a file with the same name, the generator name, including namespace, will be prepended to the name of the files to ensure there are no conflicts.
```

Comment:
The conflict thing is interesting, because practically we won't know about a collision until the second generator tries to write to the file. We _could_ go back and rename the originally produced file, or more likely only rename the second copy meaning which ever generator gets there first wins. We might consider issuing a warning at the same time.

Reply:
1. writing to a disparate location than the name asked for seems like asking for trouble (esp. if non-deterministic).
2. it feels like we can be deterministic here by running the generators up through the AddFile step, thus knowing the names of the things they want to generate.   we could then generate the final names just fine.  we can then actually invoke the writing step after that point.